### DISCUSSION TITLE
Autocomplete in editors?

### DISCUSSION THREAD

**Discussion by ColonelPhantom** [janet-lang/janet#1089] [Q&A]
I noticed that in a REPL, Janet supports autocomplete. (the one on [janet.guide](janet.guide) even shows documentation!)

 I was wondering if it's possible to get the same when working in an editor, like Emacs, Neovim or VSCode.

Maybe I'm just not too used to REPL-driven development yet? Although I do use Conjure or the VSCode plugin to send code to a running program sometimes.

**Comment by sogaiu** [comment]
I don't know how to get the same results.  I see there's some mention of asynchronous completion in the Conjure README.  I think there's a discord server for Conjure.  May be someone there can shed light on whether Janet support might be practical.

FWIW, the built-in janet repl also shows documentation via Ctrl-G, e,g. assume a repl is in the following state:

```
$ janet
Janet 1.27.0-bef5bd72 linux/x64/gcc - '(doc)' for help
repl:1:> array/clear
```

At that point if I enter Ctrl-G, I see the following output:

```
$ janet
Janet 1.27.0-bef5bd72 linux/x64/gcc - '(doc)' for help
repl:1:> array/clear


    cfunction
    src/core/array.c on line 341, column 1

    (array/clear arr)
    
    Empties an array, setting it's count to 0 but does not free the
    backing capacity. Returns the modified array.
```

I don't know what people do typically for Neovim / Vim, but in Emacs I tend to complete on the content of applicable buffers using `M-/` (dabbrev-expand).  That's pretty helpful when relevant buffers already have a fair bit of content.  It's not that great when one's buffers are sparse and/or the name of the identifier one wants is not there somewhere already.  Opening `boot.janet` in a buffer can help though :)

May be there's a feature like that in other editors too.  

I think someone was working on [an LSP server for Janet](https://github.com/JohnDoneth/janet-language-server).  I believe typical LSP approaches are not likely to give as many idenitifers as the built-in janet repl's approach because they tend to use static analysis of code and that's not likely to surface all relevant identifiers.  That doesn't mean such an approach doesn't provide value though.

**Reply by ColonelPhantom** [reply]
Vim has Ctrl-N for 'dumb keyword completion'. Opening `boot.janet` sounds like a pretty clever hack actually!

I'm looking at how it works in Conjure and I guess it can ask the REPL for a list of completions, and even documentation. Maybe that's something that I could look into, although it seems like there's no support for "context" which I guess is necessary to adapt completions to the module I am currently editing?

**Reply by sogaiu** [reply]
Since Conjure can establish a connection to a running Janet process, with that in place, if some `import` form has been evaluated, it seems possible that the public names from the corresponding module could be available because they would be part of a suitable environment.  No clue what level of support exists for this currently in Conjure though.

Until `import`ing or similar occurs, unless some other means (e.g. static analysis) is employed, it seems like the available identifiers would tend to not be comprehensive.  Though even with static analysis, although one might be able to learn more identifiers, I'm not sure one would in general be able to determine a comprehensive list.  AFAIU, that tends to be the nature of working with significantly dynamic languages.

**Reply by sogaiu** [reply]
> although it seems like there's no support for "context" which I guess is necessary to adapt completions to the module I am currently editing?

That sounds right.

I don't know if that's a limitation that can be overcome.

For reference, I think Conjure's support for "context" in Clojure relies on there being namespaces which one can switch among at will.  I don't think Janet has quite the same thing.
